aten/src/ATen/core/TensorImpl.h

```diff
@@ -863,6 +869,33 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
    */
   bool is_variable() const { return is_variable_; };
 
+  /**
+   * Set whether a tensor allows size or storage changes.
+   */
+  void set_allow_size_or_storage_change(bool value) {
+    allow_size_or_storage_change_ = value;
+  }
+
+  /**
+   * True if a tensor allows size or storage changes.
+   */
+  bool allow_size_or_storage_change() const {
+    return allow_size_or_storage_change_;
+  }
+
+  // NOTE: `shallow_copy_and_detach()` does not copy the AutogradMeta pointer
+  // because it requires unique ownership.
+  virtual c10::intrusive_ptr<TensorImpl> shallow_copy_and_detach() const {
+    auto impl = c10::make_intrusive<TensorImpl>(Storage(storage()), type_id(), is_variable());
+    impl->set_sizes_and_strides(sizes(), strides());
+    impl->storage_offset_ = storage_offset();
+    impl->is_wrapped_number_ = is_wrapped_number();
+    impl->reserved_ = reserved_;
+    impl->refresh_numel();
+    impl->refresh_contiguous();
```

Comment:
There seem to be a lot of places where we do the following:
```cpp
// In VariableType.cpp
Tensor VariableType::transpose(const Tensor & self, int64_t dim0, int64_t dim1) const {
    ... 
    auto result = as_view(self, baseType->transpose(self_, dim0, dim1), true);
    return result;
}
```

`as_view()` internally calls `make_variable_view()`, which calls `shallow_copy_and_detach()` to make a shallow copy of the TensorImpl. If we set `allow_size_or_storage_change` to false by default in `shallow_copy_and_detach()`, it would mean that the following use case would break:
```python
a = torch.DoubleTensor(1, 2)
b = a.transpose(0, 1)
b.resize_(2, 2)  # fails with "set_sizes_contiguous is not allowed"
```

I feel that we should only set `allow_size_or_storage_change` to false in user-facing APIs (i.e. `.data` and `.detach()`) to prevent misuse, because internally we might want more flexibility regarding whether we allow changing the view variable's shape/storage.

Reply:
ok, that sounds reasonable to me.